src/lua/mod.rs

```diff
@@ -0,0 +1,283 @@
+        Mutex::new(tx)
+    };
+
+    /// Receives data back from the lua thread
+    /// This should only be accessed by the lua thread itself.
+    static ref RECEIVER: Mutex<Receiver<LuaResponse>> = {
+        let (tx, rx) = channel::<LuaResponse>();
+        Mutex::new(rx)
+    };
+
+    /// Whether the lua thread is currently running
+    pub static ref RUNNING: RwLock<bool> = RwLock::new(false);
+}
+
+/// Messages sent to the lua thread
+pub enum LuaQuery {
+    /// Halt the lua thread
+    Terminate,
+    // Restart the lua thread
+    Restart,
+    /// Execute a string
+    Execute(String),
+    /// Execute a file
+    ExecuteFile(String),
+    /// Get a variable
+    GetVariable(String),
+    /// Set a value
+    SetValue {
+        name: Box<::std::borrow::Borrow<str> + Sized>,
+        val: Box<hlua::Push<&'static mut Lua<'static>> + Sized>
+    },
+    /// Create a new array
+    EmptyArray(String),
+    /// Message to ping the lua thread
+    Ping,
+    /// Unused send type
+    Unused,
+}
+
+/// Messages received from lua thread
+pub enum LuaResponse {
+    /// Lua variable obtained
+    Variable(Option<AnyLuaValue>),
+    /// Lua error
+    Error(hlua::LuaError),
+    /// A function is returned
+    Function(hlua::functions_read::LuaFunction<String>),
+    /// Pong response from lua ping
+    Pong,
+    /// Unused response type
+    Unused,
+}
+
+unsafe impl Send for LuaQuery { }
+unsafe impl Send for LuaResponse { }
+unsafe impl Sync for LuaQuery { }
+unsafe impl Sync for LuaResponse { }
+
+/// Whether the lua thread is currently available
+pub fn thread_running() -> bool {
+    *RUNNING.read().unwrap()
+}
+
+/// Errors which may arise from attempting
+/// to sending a message to the lua thread.
+pub enum LuaSendError {
+    ThreadClosed,
+    Sender
+}
+
+/// Attemps to send a LuaQuery to the lua thread.
+pub fn try_send(query: LuaQuery) -> Result<(), LuaSendError> {
+    if !thread_running() { Err(LuaSendError::ThreadClosed) }
+    else {
+        match SENDER.lock().unwrap().send(query) {
+            Ok(_) => Ok(()),
+            Err(_) => Err(LuaSendError::Sender)
+        }
+    }
+}
+
+/// Sends a value to the lua thread.
+///
+/// # Panics
+/// * If the lua thread is currently not running (check `lua::thread_running()`)
+/// * If the sender was unable to send, which should only happen
+/// if the lua thread isn't running.
+pub fn send(query: LuaQuery) {
+    if !thread_running() {
+        panic!("lua: Attempted to message the lua thread while it is not active!");
+    }
+    SENDER.lock().unwrap().send(query).unwrap();
+}
+
+/// Initialize the lua thread
+pub fn init() {
+    trace!("Initializing...");
+    let (query_tx, query_rx) = channel::<LuaQuery>();
+    let (answer_tx, answer_rx) = channel::<LuaResponse>();
+    {
+        let mut sender = SENDER.lock().unwrap();
+        let mut receiver = RECEIVER.lock().unwrap();
+
+        *sender = query_tx;
+        *receiver = answer_rx;
+    }
```

Comment:
So there are four messengers:

1. Lua thread->send
2. Lua thread<- receive
3. outside->send
4. outside<-receive

The outside ones are declared as lazy statics, and the lua thread ones are moved into the thread.

> `let sender = answer_tx;`
> `let receiver = query_rx;`

However, the lazy static macro requires a thing to be initialized. One way to do it is to have them be `Option<Sender>` and `Option<Receiver>`, and ~~~unwrap~~~ `match` the values in `try_send`. I chose to be lazy and initialize them to one end of a channel (and immediately drop the other half) before the thread's `init()` fixed them.

I'll change them to `Option` types.